core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala

```diff
@@ -204,35 +209,43 @@ private[spark] class CoarseMesosSchedulerBackend(
 
       for (offer <- offers) {
         val slaveId = offer.getSlaveId.toString
-        val mem = getResource(offer.getResourcesList, "mem")
-        val cpus = getResource(offer.getResourcesList, "cpus").toInt
-        if (totalCoresAcquired < maxCores &&
-            mem >= MemoryUtils.calculateTotalMemory(sc) &&
-            cpus >= 1 &&
+        var totalMem = getResource(offer.getResourcesList, "mem")
+        var totalCpus = getResource(offer.getResourcesList, "cpus").toInt
+
+        val tasks = new ArrayBuffer[MesosTaskInfo]
+
+        var executorCount = slaveIdsWithExecutors.getOrElse(slaveId, 0)
+        // Launch as many executors that the resources are available and less
+        // than the configured max executors per slave.
+        while (totalCoresAcquired < maxCores &&
+            totalMem >= memRequired &&
+            totalCpus >= 1 &&
             failuresBySlaveId.getOrElse(slaveId, 0) < MAX_SLAVE_FAILURES &&
-            !slaveIdsWithExecutors.contains(slaveId)) {
-          // Launch an executor on the slave
-          val cpusToUse = math.min(cpus, maxCores - totalCoresAcquired)
+            executorCount < maxExecutorsPerSlave) {
+          val cpusToUse =
+            math.min(maxCpusPerExecutor, math.min(totalCpus, maxCores - totalCoresAcquired))
           totalCoresAcquired += cpusToUse
+          totalCpus -= cpusToUse
+          totalMem -= memRequired
           val taskId = newMesosTaskId()
           taskIdToSlaveId(taskId) = slaveId
-          slaveIdsWithExecutors += slaveId
+          executorCount += 1
+          slaveIdsWithExecutors(slaveId) = executorCount
           coresByTaskId(taskId) = cpusToUse
-          val task = MesosTaskInfo.newBuilder()
+          tasks.add(MesosTaskInfo.newBuilder()
             .setTaskId(TaskID.newBuilder().setValue(taskId.toString).build())
             .setSlaveId(offer.getSlaveId)
             .setCommand(createCommand(offer, cpusToUse + extraCoresPerSlave))
             .setName("Task " + taskId)
             .addResources(createResource("cpus", cpusToUse))
-            .addResources(createResource("mem",
-              MemoryUtils.calculateTotalMemory(sc)))
-            .build()
-          d.launchTasks(
-            Collections.singleton(offer.getId),  Collections.singletonList(task), filters)
+            .addResources(createResource("mem", memRequired))
+            .build())
+        }
+
+        if (tasks.isEmpty) {
+          d.declineOffer(offer.getId)
```

Comment:
This call is basically equivalent to the old code:
-          d.launchTasks(
-            Collections.singleton(offer.getId), Collections.emptyList[MesosTaskInfo](), filters)

It's just that it's much more obvious that we're declining offers, and not trying to launch empty set of tasks.